src/init.ml

```diff
@@ -0,0 +1,198 @@
+open! Stdune
+open! Import
+
+module Project = struct
+  module Kind = struct
+    type t =
+      | Library
+      | Executable
+
+    let to_string = function
+      | Library -> "library"
+      | Executable -> "executable"
+  end
+end
+
+module File = struct
+  type t =
+    { path: Path.t
+    ; name: string
+    ; content: string }
+
+  let make path name content =
+    {path; name; content}
+
+  let full_path {path; name; _} =
+    Path.(append path (of_string name))
+
+  let create file =
+    let path = full_path file in
+    Io.write_file ~binary:false path file.content
+end
+
+module type InitContext = sig
+  val kind : Project.Kind.t
+  val pname : string
+  val dir : Path.t
+end
+
+module type ProjectDir = sig
+  val dir : Path.t
+  val files : File.t list
+end
+
+module Proj (C : InitContext) : ProjectDir = struct
+  let dir = C.dir
+  let make_file = File.make C.dir
+
+  let dune_project =
+    let lang_stanza = Dune_project.lang_stanza () in
+    let name = Dune_project.filename in
+    let content = Printf.sprintf "\
+%s
+(name %s)
+(version 0.1)
+" lang_stanza C.pname
+    in
+    make_file name content
+
+  let package_opam =
+    let name = Printf.sprintf "%s.opam" C.pname in
+    make_file name ""
+
+  let files = [dune_project; package_opam]
+end
+
+module Lib (C : InitContext) : ProjectDir = struct
+  let dir = Path.(append C.dir (of_string "src"))
+  let make_file = File.make dir
+
+  let lib_dune =
+    let name = "dune" in
+    let content = Printf.sprintf "\
+;; Uncomment stanzas to include their functionality
+(library
+  (name %s)
+  (public_name %s)
+  ;; (inline_tests)
+  ;; (libraries core)
+  ;; (preprocess (pps ppx_inline_test))
+)
+" C.pname C.pname
+    in
+    make_file name content
+
+  let lib_ml =
+    let name = "lib.ml" in
+    let content = {|let hello name = Printf.sprintf "Hello, %s" name
```

Comment:
An empty file seems better to me.

Reply:
I'll argue this point a bit, to be sure we consider the alternatives and their effect on the ecosystem. 

imo, the advantage of having a little placeholder code is that newcomers are thus given something in place to tweak and work with. E.g., with these place-holders, a brand new user can run

```
dune init my_project --bin
cd ./my_project
dune exec my_project
# Hello, project my_project
```

and have a working program that prints output in under a minutes. They can then tweak the code a bit to confirm things work as expected, and build from there.  The advantages of smoother on boarding and greater accessibility seem self-evident.

On the other hand, the cost to experienced users is negligible: they need only delete the lines they don't want. Moreover, once custom templates are implemented in subsequent additions, we can provide an easy override, so experienced user can generate blank files, or no files, or whatever. But it seems to we should set the default to be something useful for inexperienced users.

Finally, if we opt for the file to be empty, then why not just have no file at all?

Reply:
>  But it seems to we should set the default to be something useful for inexperienced users.

I disagree. We should optimize the user experience for those have invoked the tool for more than a few times. Users do not stay beginners for long and these hints will quickly become obsolete.

> Finally, if we opt for the file to be empty, then why not just have no file at all?

That's probably best. Since we're creating a wrapped library, we can't really predicate the module name the user wants anyway. Perhaps we can have a module name to specify this?

For executables of course, we should keep the empty file.

I suppose our main disagreement comes from the philosophy of this new command. I expect this command to be useful in helping all users quickly write stanza boilerplate, rather than just a tool to help beginners. I'd be interested in coming up with a compromise that could satisfy both experienced and beginning users.

Reply:
I think that's a spot on articulation of the grounds for the disagreement. I think I had these two purposes both in mind, and was not keeping them adequately separated. On reflection, I agree we should separate out these concerns.

If you agree, I'll keep this PR and immediate future work focused on the general purpose utility. Once we have the tooling built out to reduce stanza boilerplate and facilitate composable (likely configurable) generation of common stanzas, we can use it to make a very nice, dedicated beginners tool. The latter can be under its own command, perhaps as `dune starter`, `dune wizard`, or something like that. Hows that sound?